File: LanguageLearningApp.UI/Services/UserService.cs
using LanguageLearningApp.Data.Context;
using LanguageLearningApp.UI.Clients;
using Microsoft.AspNetCore.Http;
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using System.Web;
using System.IdentityModel.Tokens.Jwt;
using System.Net.Http.Headers;
using Azure;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;

public class UserService : ControllerBase
{
    private readonly HttpClient _httpClient;
    private readonly IHttpContextAccessor _httpContextAccessor;
    public User User { get; private set; }

    public List<string> ErrorMessages { get; private set; } = new List<string>();

    public UserService(UserHttpClient userHttpClient, IHttpContextAccessor httpContextAccessor)
    {
        _httpClient = userHttpClient.HttpClient;
        _httpContextAccessor = httpContextAccessor;
    }

    public async Task<bool> RegisterUserAsync(UserRegistrationDTO model)
    {
        try
        {
            // Clear previous error messages
            ErrorMessages.Clear();


            if (string.IsNullOrEmpty(model.Username) || string.IsNullOrEmpty(model.Email) || string.IsNullOrEmpty(model.Password))
            {
                ErrorMessages.Add("Please provide valid values for all required fields.");
                return false; // Registration failed due to empty fields
            }

            HttpResponseMessage response = await _httpClient.PostAsJsonAsync("user/register", model);
            if (response.IsSuccessStatusCode)
            {
                return true; // Registration successful
            }
            else if (response.StatusCode == System.Net.HttpStatusCode.BadRequest)
            {
                // Parse error response and set error messages
                var errorResponse = await response.Content.ReadAsStringAsync();
                var errorDictionary = JsonSerializer.Deserialize<Dictionary<string, List<string>>>(errorResponse);

                if (errorDictionary != null && errorDictionary.Count > 0)
                {
                    foreach (var errors in errorDictionary.Values)
                    {
                        ErrorMessages.AddRange(errors);
                    }
                }
            }

            return false; // Registration failed due to server error
        }
        catch (Exception ex)
        {
            ErrorMessages.Add("Failed to register user: " + ex.Message);
            return false; // Registration failed due to exception
        }
    }

    public string GetRefreshToken()
    {
		var context = _httpContextAccessor.HttpContext;
		return context?.Request.Cookies["accessToken"];
	}
    public async Task<string> Initialize()
    {
        var refreshToken = GetRefreshToken();
        if (string.IsNullOrEmpty(refreshToken))
        {
            return "Refresh token is missing or invalid.";
        }

        // Set the Authorization header with the refresh token
        _httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", refreshToken);

        HttpResponseMessage response = await _httpClient.GetAsync("User/email");

        if (response.IsSuccessStatusCode)
        {
            string userInfo = await response.Content.ReadAsStringAsync();
            return userInfo;
        }
        else
        {
            return "Failed to retrieve user information. HTTP status code: " + response.StatusCode;
        }
    }

    public async Task<bool> AuthenticateUserAsync(UserLoginDTO model)
    {
        try
        {
            ErrorMessages.Clear();

            if (string.IsNullOrEmpty(model.Username) || string.IsNullOrEmpty(model.Password))
            {
                ErrorMessages.Add("Please provide both username/email and password.");
                return false; // Authentication failed due to empty fields
            }

            HttpResponseMessage response = await _httpClient.PostAsJsonAsync("user/login", model);
            if (response.IsSuccessStatusCode)
            {
                return true;
            }
            else if (response.StatusCode == System.Net.HttpStatusCode.Unauthorized)
            {
                ErrorMessages.Add("Invalid username or password.");
            }
            else if (response.StatusCode == System.Net.HttpStatusCode.BadRequest)
            {
                // Parse error response and set error messages
                var errorResponse = await response.Content.ReadAsStringAsync();
                var errorDictionary = JsonSerializer.Deserialize<Dictionary<string, List<string>>>(errorResponse);

                if (errorDictionary != null && errorDictionary.Count > 0)
                {
                    foreach (var errors in errorDictionary.Values)
                    {
                        ErrorMessages.AddRange(errors);
                    }
                }
            }

            return false; // Authentication failed
        }
        catch (Exception ex)
        {
            ErrorMessages.Add("Failed to authenticate user: " + ex.Message);
            return false;
        }
    }
}

File: languagelearningapp.ui/src/services/userService.js
import React, { useEffect } from 'react';
import Cookies from 'js-cookie';

export class userService {
    constructor(baseUrl) {
        this.url = baseUrl;
    }

    #getAccessTokenFromCookies() {
        return Cookies.get('accessToken');
    }

    async #_myFetch(url, method = null, body = null) {
        try {
            method ??= 'GET';
            const accessToken = this.#getAccessTokenFromCookies();
            const headers = {
                'Content-Type': 'application/json',
            };
            if (accessToken) {
                headers['Authorization'] = `Bearer ${accessToken}`;
            }
            let res = await fetch(url, {
                credentials: 'include',
                method: method,
                headers: headers,
                body: body ? JSON.stringify(body) : null
            });

            if (!res.ok) {
                let errorData;
                try {
                    errorData = await res.json();
                } catch (err) {
                    errorData = { message: 'Unknown error' };
                }
                throw { response: { status: res.status, data: errorData } };
            }

            console.log(`\n${method} Request successful @ ${url}`);

            // Ensure response is not empty before attempting to parse as JSON
            if (res.status === 204 || res.headers.get('Content-Length') === '0') {
                return null;
            }

            const contentType = res.headers.get('Content-Type');
            if (contentType && contentType.indexOf('application/json') !== -1) {
                let data = await res.json();
                return data;
            } else {
                return null;
            }
        } catch (err) {
            console.error(`Failed to receive data from server:`, err);
            // throw err;
        }
    }

    async register(username, email, password) {
        const url = `${this.url}/register`;
        const body = {
            username: username,
            email: email,
            password: password
        };
        
        return await this.#_myFetch(url, 'POST', body);
    }

    async login(username, password) {
        const url = `${this.url}/login`;
        const body = {
            username: username,
            password: password
        };
        
        return await this.#_myFetch(url, 'POST', body);
    }

    userIsLoggedIn(){
        if(Cookies.get('accessToken')) {
            return true;
        }
        return false;
    }

    async getUserInfo() {
        const url = `${this.url}/email`;
        
        const userInfo = await this.#_myFetch(url, 'GET');

        return userInfo
    }
}